### PR TITLE
SNOW-1296607: Reduce merge gates to a minimum set

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   For precommit test, we will only have 4 required gates:
precommit test / Test py-macos-latest-3.9-aws (doctest)
precommit test / Test py-windows-latest-64-cores-3.11-gcp
precommit test / Test py-ubuntu-latest-64-cores-3.9-aws 
precommit test / Test py-ubuntu-latest-64-cores-3.10-azure 

we will also have local testing gates, and one cte merge gates. No pandas and disable sql simplifier test have been moved to daily github actions.

### CODE REVIEW

**Comment by sfc-gh-yzou** [comment]
@sfc-gh-jdu there seems some test fails in this pr as well as the daily precommit job, can you take a look?

File: .github/workflows/precommit.yml [hunk 1]
```diff
@@ -119,7 +151,7 @@ jobs:
         run: python -m pip install -U setuptools pip wheel
       - name: Install tox
         run: python -m pip install tox
-      - if: ${{ contains('macos', matrix.os.download_name) }}
+      - if: ${{ matrix.os == 'macos-latest' && matrix.cloud-provider == 'aws' }}
```
Comment (sfc-gh-azhan) [inline .github/workflows/precommit.yml:154]: add a comment for this decision

Comment (sfc-gh-jdu) [reply .github/workflows/precommit.yml:155]: added

File: .github/workflows/precommit.yml [hunk 2]
```diff
@@ -130,182 +163,15 @@ jobs:
...
-        run: python -m tox -e "py${PYTHON_VERSION}-doctest-notudf-ci"
-        env:
-          PYTHON_VERSION: ${{ matrix.python-version }}
-          cloud_provider: ${{ matrix.cloud-provider }}
-          PYTEST_ADDOPTS: --color=yes --tb=short --disable_sql_simplifier
-          TOX_PARALLEL_NO_SPINNER: 1
-        shell: bash
-      - name: Run tests (excluding doctests)
+      # do not run other tests for macos on aws
+      - if: ${{ !(matrix.os == 'macos-latest' && matrix.cloud-provider == 'aws') }}
```
Comment (sfc-gh-azhan) [inline .github/workflows/precommit.yml:167]: We can just skip macos here right?

Comment (sfc-gh-jdu) [reply .github/workflows/precommit.yml:167]: You mean we don't need to have the check for aws now?

Comment (sfc-gh-azhan) [reply .github/workflows/precommit.yml:167]: I guess right now no macos should run in parallel in merge gate right?

**Review by sfc-gh-azhan — APPROVED** [review]

File: .github/workflows/precommit.yml [hunk 3]
```diff
@@ -130,182 +163,15 @@ jobs:
...
-        name: Run doctests
-        run: python -m tox -e "py${PYTHON_VERSION}-doctest-notudf-ci"
-        env:
-          PYTHON_VERSION: ${{ matrix.python-version }}
-          cloud_provider: ${{ matrix.cloud-provider }}
-          PYTEST_ADDOPTS: --color=yes --tb=short --disable_sql_simplifier
-          TOX_PARALLEL_NO_SPINNER: 1
-        shell: bash
-      - name: Run tests (excluding doctests)
+      # do not run other tests for macos on aws
```
Comment (sfc-gh-azhan) [inline .github/workflows/precommit.yml:166]:
```suggestion
      # do not run other tests for macos
```

File: .github/workflows/precommit.yml [hunk 4]
```diff
@@ -77,21 +77,53 @@ jobs:
...
-        os:
-          - image_name: ubuntu-latest
-            download_name: linux
-          - image_name: macos-latest
-            download_name: macos
-          - image_name: windows-latest
-            download_name: windows
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        os: [macos-latest, windows-latest-64-cores, ubuntu-latest-64-cores]
+        python-version: ["3.9", "3.10", "3.11"]
```
Comment (sfc-gh-jrose) [inline .github/workflows/precommit.yml:87]: python 3.8 isn't eol until October. Should we keep it in the test matrix until then?

Comment (sfc-gh-azhan) [reply .github/workflows/precommit.yml:87]: 3.8 is still tested in daily test job.

Comment (sfc-gh-jdu) [reply .github/workflows/precommit.yml:87]: We still have 3.8 test running every day https://github.com/snowflakedb/snowpark-python/blob/04b84a36e5405793fa96f7d2bb73a608424b2e0d/.github/workflows/daily_precommit.yml#L95

**Review by sfc-gh-nkrishna — APPROVED** [review]

**Review by sfc-gh-jrose — APPROVED** [review]

**Pull request merged**